test/unix/main.ml

```diff
@@ -4,234 +4,242 @@ let ( // ) = Filename.concat
+    Index.close rw;
+    let rw = clone ~readonly:false in
+    check_entries_present rw tbl
+
+  let find_absent () =
+    let { Context.rw; tbl; clone } = Context.full_index () in
+    Index.close rw;
+    let rw = clone ~readonly:false in
+    test_find_absent rw tbl
+
+  let replace () =
+    let { Context.rw; clone; _ } = Context.full_index ~size:5 () in
+    Index.close rw;
+    let rw = clone ~readonly:false in
+    test_replace rw
+
+  let tests =
+    [
+      ("find (present)", `Quick, find_present);
+      ("find (absent)", `Quick, find_absent);
+      ("replace", `Quick, replace);
+    ]
+end
+
+(* Tests of read-only indices *)
+module Readonly = struct
+  let readonly () =
+    let name, w = Context.empty_index () in
+    let r = Index.v ~fresh:false ~readonly:true ~log_size name in
+    Hashtbl.iter (fun k v -> Index.replace w k v) main.tbl;
+    Index.flush w;
+    check_entries_present r main.tbl
+
+  let readonly_clear () =
+    let { Context.rw; tbl; clone } = Context.full_index () in
+    let r = clone ~readonly:true in
+    check_entries_present r tbl;
+    Index.clear rw;
+    Hashtbl.iter
+      (fun k _ ->
+        Alcotest.check_raises (Printf.sprintf "Found %s key after clearing." k)
+          Not_found (fun () -> ignore (Index.find r k)))
+      tbl
+
+  let tests =
+    [ ("add", `Quick, readonly); ("read after clear", `Quick, readonly_clear) ]
+end
+
+(* Tests of {Index.close} *)
+module Close = struct
+  let close_reopen_rw () =
+    let { Context.rw; tbl; clone } = Context.full_index () in
+    Index.close rw;
+    let w = clone ~readonly:false in
+    check_entries_present w tbl;
+    Index.close w
+
+  let open_readonly_close_rw () =
+    let { Context.rw; tbl; clone } = Context.full_index () in
+    let ro = clone ~readonly:true in
+    Index.close rw;
+    check_entries_present ro tbl;
+    Index.close ro
+
+  let close_reopen_readonly () =
+    let w = Index.v ~fresh:true ~readonly:false ~log_size (root // "test3") in
+    Hashtbl.iter (fun k v -> Index.replace w k v) main.tbl;
+    Index.close w;
+    let r = Index.v ~fresh:false ~readonly:true ~log_size (root // "test3") in
+    check_entries_present r main.tbl;
+    Index.close r
+
+  let test_read_after_close t k =
+    check_entries_present t main.tbl;
+    Index.close t;
+    Alcotest.check_raises "Read after close raises Not_found" Not_found
+      (fun () -> ignore (Index.find t k))
+
+  let test_read_after_close_readonly t k tbl =
+    check_entries_present t tbl;
+    Index.close t;
+    let exn = Unix.Unix_error (Unix.EBADF, "read", "") in
+    Alcotest.check_raises "Cannot read in readonly index after close." exn
+      (fun () -> ignore (Index.find t k))
+
+  let fail_read_after_close () =
+    let w = Index.v ~fresh:true ~readonly:false ~log_size (root // "test4") in
+    Hashtbl.iter (fun k v -> Index.replace w k v) main.tbl;
+    let k = Key.v () in
+    let v = Value.v () in
+    Index.replace w k v;
```

Comment:
You need to add the pair `k,v` to the hash table too.

Reply:
Good catch; we should iterate in both directions to catch this.